src/Pos/Context/Holder.hs

```diff
@@ -0,0 +1,125 @@
+    } deriving ( Functor
+               , Applicative
+               , Monad
+               , MonadTrans
+               , MonadThrow
+               , MonadCatch
+               , MonadMask
+               , MonadIO
+               , MonadFail
+               , HasLoggerName
+               , CanLog
+               , MonadSlotsData
+               , MonadSlots
+               , MonadTxpMem
+               , MonadFix
+               , MonadDB
+               , MonadDBLimits
+               )
+
+-- | Run 'ContextHolder' action.
+runContextHolder :: NodeContext ssc -> ContextHolder ssc m a -> m a
+runContextHolder ctx = flip runReaderT ctx . getContextHolder
+
+instance Monad m => WrappedM (ContextHolder ssc m) where
+    type UnwrappedM (ContextHolder ssc m) = ReaderT (NodeContext ssc) m
+    _WrappedM = iso getContextHolder ContextHolder
+
+instance MonadBase IO m => MonadBase IO (ContextHolder ssc m) where
+    liftBase = lift . liftBase
+
+type instance ThreadId (ContextHolder ssc m) = ThreadId m
+type instance Promise (ContextHolder ssc m) = Promise m
+type instance SharedAtomicT (ContextHolder ssc m) = SharedAtomicT m
+type instance Counter (ContextHolder ssc m) = Counter m
+type instance Distribution (ContextHolder ssc m) = Distribution m
+type instance SharedExclusiveT (ContextHolder ssc m) = SharedExclusiveT m
+type instance Gauge (ContextHolder ssc m) = Gauge m
+type instance ChannelT (ContextHolder ssc m) = ChannelT m
+
+instance ( Mockable d m
+         , MFunctor' d (ReaderT (NodeContext ssc) m) m
+         , MFunctor' d (ContextHolder ssc m) (ReaderT (NodeContext ssc) m)
+         ) => Mockable d (ContextHolder ssc m) where
+    liftMockable = liftMockableWrappedM
+
+instance Monad m => WithNodeContext ssc (ContextHolder ssc m) where
+    getNodeContext = ContextHolder ask
+
+instance (MonadIO m, Mockable Catch m, WithLogger m) => MonadJL (ContextHolder ssc m) where
+    jlLog ev = ContextHolder (asks ncJLFile) >>= maybe (pure ()) doLog
+      where
+        doLog logFileMV =
+          (liftIO . withMVar logFileMV $ flip appendJL ev)
+            `catchAll` \e -> logWarning $ sformat ("Can't write to json log: " % shown) e
+
+instance Monad m => MonadReportingMem (ContextHolder ssc m) where
+    askReportingMem =
+        ContextHolder $
+            asks (ReportingContext . npReportServers . ncNodeParams)
+
+instance Monad m => MonadDhtMem (ContextHolder ssc m) where
+    askDhtMem =
+        ContextHolder $ asks (DhtContext .
+                              bpKademliaDump .
+                              npBaseParams .
+                              ncNodeParams)
+instance Monad m => MonadRelayMem (ContextHolder ssc m) where
+    askRelayMem =
+        ContextHolder
+            (RelayContext
+                <$> asks (npPropagation . ncNodeParams)
+                <*> asks ncInvPropagationQueue
+            )
+
+instance Monad m => MonadShutdownMem (ContextHolder ssc m) where
```

Comment:
⚠️ All these monads aren't necessary: if all they do is read, then `MonadReader` and a `HasFoo` class is sufficient. New monads should ideally only be introduced because they have _new behavior_.